### PR TITLE
optimize consensus signature number and verify header sign

### PR DESCRIPTION
**Pull request by xiemylogos**
The number of optimized consensus node packing block signatures must be greater than or equal to 5/7.

@Honglei-Cong  @luodanwg @laizy

### CODE REVIEW

**Review by Honglei-Cong — DISMISSED** [review]

File: consensus/vbft/service.go [hunk 1]
```diff
@@ -1963,13 +1963,17 @@ func (self *Server) fastForwardBlock(block *Block) error {
 	}
 	if self.GetCurrentBlockNo() == block.getBlockNum() {
 		// block from peer syncer, there should only one candidate block
-		return self.sealBlock(block, false)
+		flag := false
+		if len(block.Block.Header.SigData) == 0 {
+			flag = true
+		}
+		return self.sealBlock(block, false, flag)
```
Comment (Honglei-Cong) [inline consensus/vbft/service.go:1977]: return self.sealBlock(block, false, len(block.Block.Header.SigData) == 0） ？

or

toAddSignatures = len(block.Block.Header.SigData) == 0
return self.sealBlock(block, false, toAddSignatures) ??

Comment (Honglei-Cong) [inline consensus/vbft/service.go:1977]: return self.sealBlock(block, false, len(block.Block.Header.SigData) == 0） ？

or

toAddSignatures = len(block.Block.Header.SigData) == 0
return self.sealBlock(block, false, toAddSignatures) ??

File: consensus/vbft/block_pool.go [hunk 2]
```diff
@@ -314,22 +314,22 @@ func (pool *BlockPool) endorseDone(blkNum uint32, C uint32) (uint32, bool, bool)
 		return math.MaxUint32, false, false
 	}
 
-	if uint32(len(candidate.EndorseSigs)) < C+1 {
+	if uint32(len(candidate.EndorseSigs)) < 2*C {
```
Comment (Honglei-Cong) [inline consensus/vbft/block_pool.go:317]: why change this?

File: consensus/vbft/node_utils.go [hunk 3]
```diff
@@ -233,13 +233,29 @@ func calcParticipantPeers(cfg *BlockParticipantConfig, chain *vconfig.ChainConfi
 
 	peers := make([]uint32, 0)
 	peerMap := make(map[uint32]bool)
+	proposerMap := make(map[uint32]bool)
 	var cnt uint32
 
+	if checkCalcEndorserOrCommitter(end) {
+		if len(cfg.Proposers) != 0 {
```
Comment (Honglei-Cong) [inline consensus/vbft/node_utils.go:240]: no needed 'if'

File: consensus/vbft/node_utils.go [hunk 4]
```diff
@@ -264,6 +279,14 @@ func calcParticipantPeers(cfg *BlockParticipantConfig, chain *vconfig.ChainConfi
 	return peers
 }
 
+func checkCalcEndorserOrCommitter(end int) bool {
```
Comment (Honglei-Cong) [inline consensus/vbft/node_utils.go:282]: change this function to 'isCalcProposer(end) bool' ?

File: consensus/vbft/node_utils.go [hunk 5]
```diff
@@ -264,6 +279,14 @@ func calcParticipantPeers(cfg *BlockParticipantConfig, chain *vconfig.ChainConfi
...
+	if end == vconfig.MAX_ENDORSER_COUNT+vconfig.MAX_PROPOSER_COUNT ||
```
Comment (Honglei-Cong) [inline consensus/vbft/node_utils.go:283]: just return (a == b)

File: consensus/vbft/service.go (hunk 1 above)
Comment (xiemylogos) [reply consensus/vbft/service.go:1977]: ok

Comment (xiemylogos) [reply consensus/vbft/service.go:1977]: ok

File: consensus/vbft/node_utils.go (hunk 3 above)
Comment (xiemylogos) [reply consensus/vbft/node_utils.go:240]: yes

File: consensus/vbft/node_utils.go (hunk 4 above)
Comment (xiemylogos) [reply consensus/vbft/node_utils.go:282]: ok

File: consensus/vbft/node_utils.go (hunk 5 above)
Comment (xiemylogos) [reply consensus/vbft/node_utils.go:283]: yes

File: consensus/vbft/block_pool.go [hunk 6]
```diff
@@ -322,14 +322,14 @@ func (pool *BlockPool) endorseDone(blkNum uint32, C uint32) (uint32, bool, bool)
 		for _, esig := range eSigs {
 			if esig.ForEmpty {
 				emptyEndorseCount++
-				if emptyEndorseCount > int(C) {
+				if emptyEndorseCount >= int(C) {
```
Comment (Honglei-Cong) [inline consensus/vbft/block_pool.go]: in your last change, there'll C+1 proposers, the last proposer can also be endorser.
So, there should always be at least C+1 honest endorsers.
No change need here.

Comment (xiemylogos) [reply consensus/vbft/block_pool.go]: already fixed

**Review by Honglei-Cong — APPROVED** [review]

**Pull request merged**